Database merge confirmation dialog

taminob opened the pull request:
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Explain large or complex code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" )

This PR adds a dialog allowing a user to review the changes of a merge operation.
This dialog displays the changes and allows the user to abort the merge without modifying the database.

Fixes #1152

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )

![merge dialog](https://github.com/keepassxreboot/keepassxc/assets/28907748/b0470a81-4f58-4a63-bbc0-748fc9ba1c60)

![merge aborted](https://github.com/keepassxreboot/keepassxc/assets/28907748/9645c067-1e21-47d8-bdf7-f1592d9afe3d)

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )

Manual testing and the execution of the already existing test suite `testmerge`.

 I also added a some basic tests verifying that a database will remain unmodified when only calculating the changelist and if for a merge without re-calculation only that changelist will be applied to the database.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature
- ✅ Refactor (small refactoring)

taminob commented:
This PR doesn't yet follow the contribution guidelines and contains temporary code, so don't take it too seriously.

I did a first draft to simply be able to call with a `dryRun` parameter which would disable all modifying operations.
However, I feel like this is kind of dirty and doesn't really help with the code complexity.
Thus, I tried to split off the modifying operations and be able to execute them afterwards which would also already enable a future feature like only partially merging a database (although this will require some additional logic).

I was struggeling a bit with the lifetime of the objects and if there are any guarantees defined somewhere and if e.g. the entry pointers are actually guaranteed to be not null and will remain valid.

The last major open task is to fix the merge of the deletions since the current implementation relies on the other entries being merged first - however, this change requires to not modify anything before confirmed by the user.
Let me know if you'd rather keep these changes minimal and I'll consider less invasive alternatives like merging into a temporary database.

droidmonkey commented:
Awesome MVP!!

taminob committed:
Introduce Database Merge Confirmation Dialog

Adds a dialog allowing a user to review the changes of a merge operation.
This dialog displays the changes and allows the user to abort the merge
without modifying the database.

Fixes #1152

taminob commented:
Sorry that the PR became kind of huge - I encountered some unexpected issues with the merge logic for deletions since it relied on the groups/entries already being merged before.
However, I need those merged deletions before the actual merge to display the change list.
Please read the changes carefully to avoid any data loss in the merges since I could easily have missed some edge case.

Thanks a lot to whoever wrote the extensive test suite for the merge logic, that helped a lot to debug issues and also understand the reasoning behind the merge behavior.

If you want to, I can also split this into two PRs, I already kept it apart in two commits (one for the merge logic refactoring and one for the UI).

I also left a couple of `TODO` markers in there with a few questions regarding the existing code and if that's still up-to-date.

mmahmoudian commented:
@taminob Thanks for investing time into implementing this. I have suggested some related CLI features and I've been told to suggest them here as this PR is the perfect place for this.


It would be amazing if we can have the table you implemented in GUI into CLI since the info is there and it would be just a matter for formatting (for instance in markdown table format):


```console
❯  keepassxc-cli merge --dry-run --same-credentials db-local.kdbx db-remote.kdbx
Enter password to unlock db-local.kdbx

| Action                               | what      | UUID               |
|--------------------------------------|-----------|--------------------|
| Overwriting                          | blah blah | [814h814h814h814h] |
| '_ Synchronizing from newer source   | blah blah | [814h814h814h814h] |
|   '_ Creating missing                | blah blah | [814h814h814h814h] |
|   '_ Creating missing                | blah blah | [814h814h814h814h] |

Database was not modified by merge operation.
```

Additionally, It would be extra useful to specify which file is the "newer source" (newer can be based on file modification time or internal variable in databases). In the screenshot you  posted in GUI, you also have "newer source" which is imho ambiguous.


Also (slightly off-topic), I wonder if it is feasible to use some sort of color coding. Then `--color=[auto,always,never]` can control that like other commonly use CLI tools.

As a final remark and comment/feedback on your amazing work, I would like to suggest to move the UUID to the last column of the table as to the user UUID is not so useful compared to other columns.

Thanks again for your efforts. I've been waiting for this feature for quite some time :)

taminob commented:
@mmahmoudian thanks for suggesting, I think that would be an improvement, too. I also like your mockup, although I'm not quite sure what's the meaning of the tree-like structure in the `Action` column?

However, I think that it might be better to keep further user experience improvements to a separate follow-up PR since this PR is already quite complex. I even thought about moving the entire GUI stuff to a separate PR.
Although I noticed that some of the complexity might be not even necessarily required when looking at the `--dry-run` implementation in the CLI tool (unfortunately didn't notice that before - could've saved me a lot of trouble implementing the refactoring :) ) since modifications to the `Database` object seem to be safe as long as it's not saved (and the dialog is blocking all other input anyway, so user shouldn't be able to save manually).

Do you have any opinion @droidmonkey, do you think the current refactoring is worth the risk? I'd also be willing to look into the CLI improvements, generating such a table shouldn't be too hard.

Otherwise, this PR would also be ready if you have time to review it. The failing CI is unfortunately caused by out-dated translations on `develop`.

mmahmoudian commented:
@taminob 

> what's the meaning of the tree-like structure in the Action column?

Basically this hierarchical structure is better render of what the CLI outputs at the moment. To illustrate it further, compare the screenshot and the table in this post:

https://github.com/keepassxreboot/keepassxc/issues/6264#issuecomment-1937037658

> However, I think that it might be better to keep further user experience improvements to a separate follow-up PR since this PR is already quite complex.

Sure. Definitely you know better :)

droidmonkey commented:
Generally speaking, I try to avoid refactor unless we are solving a code pain point. Looks like you'd be good to not make all this additions to Merger.h/cpp

taminob commented:
> Generally speaking, I try to avoid refactor unless we are solving a code pain point. Looks like you'd be good to not make all this additions to Merger.h/cpp

I completely understand that and agree that unnecessary refactorings can cause a lot of pain.

I looked again at the alternative which is modifying the database (or a temporary copy of it) and restore the original database if the merge was aborted.
However, I noticed that the database might get automatically reloaded when changed on disk and I'm also not sure if the database might get saved automatically in the background without the merge being confirmed.
For the first case, I think it would be possible to avoid reloading via an additional `isMergeInProgress` check in https://github.com/keepassxreboot/keepassxc/blob/a472ef8a931dd64990d7f735394e949638e880d5/src/gui/DatabaseWidget.cpp#L1838, but I'm not sure if the second requirement can be guaranteed somehow. If the merge would modify the database anyway (or the actually merged changes do not match the displayed ones), that would, in my opinion, be way worse than simply not having the confirmation dialog.

taminob commented on src/gui/MergeDialog.cpp:
```diff
@@ -0,0 +1,128 @@
+void MergeDialog::setupChangeTable()
+{
+    if (m_changes.isEmpty()) {
+        // deep copy of root group with same uuid
+        auto tmpRootGroup = m_targetDatabase->rootGroup()->clone(Entry::CloneFlag::CloneIncludeHistory,
+                                                                 Group::CloneFlag::CloneIncludeEntries);
+        Database tmpDatabase;
+        tmpDatabase.setRootGroup(tmpRootGroup);
+        m_changes = Merger(m_sourceDatabase.data(), &tmpDatabase).merge();
```
@droidmonkey while this kind works, is there a better way to clone a database? Wasn't able to find one in the current code base.
Also, this does not include the `Database::m_metadata` which doesn't leads to missing changes regarding Recycling Bin and Custom Icons - is there a reason why it doesn't implement a `clone()` function or should I implement it to do a full deep copy of a database?

droidmonkey replied:
I'd have to check but I don't think so since cloning a database is not really a thing in the program. Not saying it wouldn't be useful, but the current method you are using here is not the correct way. That would be more like transferring root group and children.

taminob replied:
But doesn't `Group::clone()` create an independent copy of the root group (the ownership of which I then transfer to the temporary database)?

The alternative to cloning the database would be to implement something like `discardModifications()` for a database - as far as I can see, that doesn't exist yet. But then we have to be absolutely sure that no one will save the database in the meantime before the merge is confirmed.

Which of these two approaches do you think fit the current design better? Or do you have another idea on how to achieve such a "temporary merge" without major changes to the merge logic?

Qwerty-Space commented:
Another (hopefully small) addition you could make to this merge window is the option to exclude entries from the merge.

Let's say we have DB foo and DB bar both with some new and outdated entries.  You might want to exclude the outdated entries when merging to avoid having to roll them back later.

taminob commented:
> Another (hopefully small) addition you could make to this merge window is the option to exclude entries from the merge.
> 
> Let's say we have DB foo and DB bar both with some new and outdated entries. You might want to exclude the outdated entries when merging to avoid having to roll them back later.

When implementing my first draft, I had this feature already in mind (although for a separate PR). However, this could require a complete rewrite of the merge logic which was rejected before (see https://github.com/keepassxreboot/keepassxc/pull/10173#issuecomment-1937085641) for valid reasons - refactoring such a complex part of the code is not without risks.
I can have a look if we could implement this without the refactoring once we have database cloning, but no promises that that will work. But it is a more complex change than it might seem because if you e.g. exclude the deletion of a Group, you need to add all the differences of the entries in the group to the merge again (which were previously removed because of the deletion).

Otherwise, I'm currently waiting on feedback on https://github.com/keepassxreboot/keepassxc/pull/10173#discussion_r1508005059 @droidmonkey - some hints on how you think the feature would fit the current architecture best would be appreciated.

mmahmoudian commented:
With the help of a friend we managed to build this PR in Guix. If you want to follow the steps:

1. put this Scheme code in a file (e.g., named `this-file.scm`):

```lisp
(use-modules (gnu packages password-utils)
	     (gnu packages crypto)
	      (guix git-download)
	     (guix packages))
(define-public keepassxc/taminob
  (package
   (inherit keepassxc)
   (source
    (origin
     (method git-fetch)
     (uri
      (git-reference
       (url "https://github.com/taminob/keepassxc.git")
       (commit "feature/1152/database-merge-review-dialog")))
     (sha256
       (base32
	"1iscw3cq3f2fv5yiw20q73kwaicmipg7zxqm3w544zkl0kdrqala"))))
   (inputs (modify-inputs (package-inputs keepassxc)
			  (append keyutils)))))
			  
keepassxc/taminob

```

2. run this Scheme file with (remember to match the file name). Then this command would first pull all the dependencies and etc. from Guix (because it inherits the package definition), then builds Keepassxc with the source of this PR, and then will run it.

```sh
$(guix build -f this-file.scm)/bin/keepassxc
```

-------

## Now my observations:

The merge window's content are hard to interpret (at least for me):

 
![Screenshot_20250212_175546](https://github.com/user-attachments/assets/5f4c622c-7ea8-47fb-bac1-2563addd4214)

and when merged, things are even more confusing (note that this is another merge so the content is different compared to previous screenshot:

![Screenshot_20250212_175300](https://github.com/user-attachments/assets/783cc5c4-ef61-45ed-a61e-8345896c0d7d)


You might ask what is confusing, and imho the following:
1. Why every time I merge, the random slug changes? is that a normal thing? is it expected? Should the user do something about it?
2. what is a "group property" and which one of them was overwritten? From what to what? I assume it is group icon, but no way to know
3. there are only the possibilities to merge or abort. what if user wants something and not other things? Cherry picking would be super handy imho

In general, @taminob did a phenomenal job in my opinion and this PR actually saved my time today in merging three versions of the same database, but there are rough edges that can be improved. For instance having backup of the DB because the merge automatically saves the DB and over-writes the DB file, and there is no "undo".

Anyways, just wanted to share my experience, and I hope it can be useful in improving and merging this PR.

taminob commented:
Hi @mmahmoudian, thanks for trying out this PR! Unfortunately I didn't have much time in the last couple of months (and probably the next ones as well), so I didn't make any progress with this feature.

Regarding your questions:

1. no idea, sorry
2. this will overwrite the properties of your target file (the source file shouldn't be modified in a merge); but I agree that the naming could be better - I simply used the existing debug output which was printed to the console already; it includes the following changes:
  - group name
  - group notes
  - group icon
  - group expiry time
  - group modification date
3. I agree that this would be an awesome feature; however, the merge code is quite complex and does not allow for such a thing; I tried to implement what you're proposing already and made some steps into the right direction, but this is an entire rewrite of the entire logic and that was rejected as part of this PR - we'll probably need to do this in a separate PR explaining why the changes are necessary; also, this is a lot more complicated since you'll additionally need to handle dependencies because what do you do if the user selects a new entry without the also newly created parent group?

I fully agree that currently there is no way around a backup before a merge - please always do that anyway when testing experimental features since data loss can always happen due to unintended side effects.

@droidmonkey maybe we can discuss how to proceed until I have some spare time to finish this feature? Do you have an idea regarding https://github.com/keepassxreboot/keepassxc/pull/10173#discussion_r1508005059?

droidmonkey commented:
I'd like to move this forward after we release 2.7.10

mmahmoudian commented:
@taminob 

Thanks for the detailed reply. I thought I just provide my two cents on some of the points you raised:

> what do you do if the user selects a new entry without the also newly created parent group?

The parent should be created automatically, unless the parent existed but modified, which in that case the user can decide if they want to merge that modification or not (but then the parent exists anyways).

Regarding the backup and giving safety net to the user, here is what we can do: 
1. Provide an "undo" by generating a json object inside the target database that indicates what is changed to what. This way if the user wants to revert the changes, we can "replay" the json but backwards.
2. The json can serve as a log file for user to know what has happened
3. Before merging happens, the keepassxc can create a backup with timestamp in the name and inform the user that they should remove it after they have made sure the merging went as expected.

The items 1 and 2 require that the merge logic have three somewhat separate modules:
1. The module that diffs and creates a json and saves it in the target db.
2. The module that presents the json in a GUI and allows the user to check and modify what to be merged (I'm just adding my wishes here about the cherrypicking😅), and modifies the json respectively and overwrites the original on in the db.
3. The module that can "play" json and apply the changes

The separation of these modules would help develop them independently as the only object passed between them is the json object.

I truly appreciate the PR. It is a game changer imho. Thank you for the time and efforts you have put into it.

taminob commented:
@mmahmoudian you bring up some great points, e.g. the backup database is probably pretty easy to implement.
However, for review purposes, I'd keep it out of this PR and will try to remember submitting another PR once I have some time to spare.

> The parent should be created automatically, unless the parent existed but modified, which in that case the user can decide if they want to merge that modification or not (but then the parent exists anyways).

That was just an example, when I thought about it the last time I came across multiple of these cases.
I don't say that it is impossible, but that we need to completely restructure the merge logic (because the current logic also relies on the merge already being in progress; meaning that if you don't perform a full merge of e.g. the groups you won't even get the diff for the new entries).

> Regarding the backup and giving safety net to the user, here is what we can do:

Maybe you can create a new issue for this and add your description to the issue?
I fear that it will get lost here and I agree with the maintainers that we should keep the PRs small for review purposes.
In order to get this PR merged at some point, I think it really should only focus on the addition of the UI (and maybe improvements in the wordings like your suggestion with changing "overwriting group properties").
Feel free to ping me in the newly created issue so that I can also add some information like commit https://github.com/keepassxreboot/keepassxc/pull/10173/commits/93f01c3e57e83d371c90619c15ae4975b1752fe8 which tried to adapt the merge logic, but is insufficient and does not cover the dependencies which I mentioned earlier.

droidmonkey commented:
I don't see any real reason to add so much complexity to the merge operation (ie the json and undo stuff). We already offer a simple way to backup your database  (Database -> Save Backup...) or you can add the same button on the merge dialog pre-commit. Merge is meant to be used between two database files that are related, just at different points in time. It's also meant to take two different databases and make them one. In either case, an undo function doesn't make any sense and a pre-commit backup is a much better and easier option.

mmahmoudian commented:
@droidmonkey the main reason I suggested json was because it provides a clear and actionable log file. As a user, I expect three things from a merge operation:
1. Clear explanation of what is merging/merged
2. Ability to cherrypick (because both DBs could have modified independently and diverged)
3. A safety net 

The third can be provided by the existing backup functionality. This PR brings the GUI of what is going to be merged but the log file of what was merged and the cherry pick is not there. It is a very very awesome step forward though and already saved me a lot of headaches (thanks @taminob ). I naively thought that when the current PR is getting the diff and showing it in a table, it practically had all the information to write them in json object too. Perhaps I'm wrong and there are difficulties I've not foreseen.

Anyways, thank you both for the discussion and for the PR. 😊

taminob commented:
> I'd like to move this forward after we release 2.7.10

@droidmonkey ping - if I see correctly, 2.7.10 has been released by now.

The main issue with this PR at the moment is how to do a temporary merge.
The options I see for this (without merge logic modifications) are described in https://github.com/keepassxreboot/keepassxc/pull/10173#discussion_r1508005059.
What would be your preferred approach that fits the design of KeepassXC best?

droidmonkey commented:
@taminob you need to allow me to push to your branch in order to rebase this PR and review it. I have a rebase ready to go but you disabled pushing.

taminob commented:
> @taminob you need to allow me to push to your branch in order to rebase this PR and review it. I have a rebase ready to go but you disabled pushing.

@droidmonkey thanks for doing the work, you should have access now.

droidmonkey commented:
I think we can easily add checkboxes to the approval dialog so that excluding certain actions is as easy as unchecking the line item. I am willing to add that change. Otherwise, this is a very welcome code addition.

Some other thoughts:
* Don't show UUID column by default, it doesn't add any information for vast majority of users
* The type of change column might be unnecessary, especially if you prefix each change string with the type of change. For example: `Modified: Added group property XXXX`, `Created`, `Deleted`, `Custom Data: Added XXXXX`
* The "What was changed" should include the output from `EntryHistoryModel::calculateHistoryModifications` which really should be moved into Entry class and take a current and previous entry as parameters.

taminob commented:
> I think we can easily add checkboxes to the approval dialog so that excluding certain actions is as easy as unchecking the line item. I am willing to add that change. Otherwise, this is a very welcome code addition.

Do you have an idea how to implement this without major changes to the merge logic? For easy stuff like "Do not add Group X" we can simply create a temporary database again and delete the group from that database (and thus prevent the merge), but e.g. selective merging of the history could be more complex.

If you have an idea on how to do this, I can also work on implementing this, otherwise I'd appreciate your help in working on this.

All your other suggestions sound great and easy to add.

droidmonkey commented:
Hmmm good point, I had it in my mind as much easier than it is. We would need to implement true transactional changes to make that possible.

taminob committed:
Add Entry::calculateDifference()

This new function contains the logic that was previously in
EntryHistoryModel::calculateHistoryModifications().
It allows the re-use to display the differences in case of a merge.

Related to #1152

taminob committed:
Add Database::clone()

In order to allow merging on a temporary database, such a temporary
database has to be created first.
For this, all metadata are copied and a deep clone of the root group
allows completely separated operations on the database objects to
perform a temporary merge.

Related to #1152

taminob commented:
I tried to implement your suggestions, maybe you can have another look at it if it is what you meant.
Instead of removing the "Type of change" column I renamed it to just "Change" and hid the "Details" column by default (although I've also rephrased most of the strings to be more user-readable).

Default:

![image](https://github.com/user-attachments/assets/178f8c53-d59e-4229-9395-eccde9a27aa6)

With a context menu on the header to hide/unhide columns:

![image](https://github.com/user-attachments/assets/978e79fc-932e-4805-9518-b28439557833)

I also added a `clone()` function to `Database` because the previous method didn't duplicate the metadata. I hope this is okay and doesn't have some unintended side effects. If it could have any, I can also move it to the `MergeDialog` to make sure that it is not universally usable.
The copy does not contain the file path and keys of the database to hopefully prevent most issues.

For now, I kept the changes to `Entry` and `Database` in separate commits, should I squash this all into one commit or keep it separate to avoid one huge commit?

taminob commented:
@droidmonkey ping - I'd like to get this PR merged at some point. Can you maybe have another look at the current state of the changes?

droidmonkey commented:
Yes i am doing PR closeouts this week.

droidmonkey committed:
Simplified code and fixed issues

* Added dry run option to Merger
* Changed behavior when actual merge differs from dry run to just output a warning to console
* Fixed KeeShare conflicting with merge operations in the middle of a merge

droidmonkey committed:
Revert "Add Database::clone()"

This reverts commit 5b41eaf0c95f2714d889de4ac61afc79d08f5ac7.

droidmonkey commented:
Thank you for the contribution, sorry for the long delay!
